Combine the changes of SOL-199 and SOL-221

asadiqbal08 opened the pull request:
@mattdrayer please look into this and update the sandbox. 
Thanks

asadiqbal08 commented:
test this please

asadiqbal08 commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -899,8 +899,7 @@ def settings_handler(request, course_key_string):
-                    # and the course does not already have an entrance exam attached...
-                    if entrance_exam_enabled and not course_entrance_exam_present:
+                    if entrance_exam_enabled:
                         # Load the default minimum score threshold from settings, then try to override it
```
@mattdrayer what is the purpose of this check  'not course_entrance_exam_present' ? 
actually i am not sure But this is not allowing to update exam grade score , of a already created entrance exam.

mattdrayer replied:
This check was put in place to avoid inserting duplicate entrance exam chapters/modules via create_entrance_exam upon form submission.  However, it could be unnecessary at this point?

asadiqbal08 replied:
Hmmm, matt i think i should revert back this condition into its original state like as >  
  if entrance_exam_enabled and not course_entrance_exam_present:
Because during creating an entrance exam , we are first cleaning out any old entrance exams in entrance_exam.py line 84.
So this condition  > if entrance_exam_enabled: -- will create a new instance of an entrance exam instead of updating the minimum passing score of a pre created exam.

mattdrayer replied:
Right -- I think we need to decouple the creation of the exam and the
setting/changing of the score.
On Feb 17, 2015 2:23 AM, "Asad" <notifications@github.com> wrote:

> In cms/djangoapps/contentstore/views/course.py
> <https://github.com/edx/edx-platform/pull/6997#discussion_r24797292>:
>
> > @@ -899,8 +899,7 @@ def settings_handler(request, course_key_string):
> >                      ee_min_score_pct = request.json.get('entrance_exam_minimum_score_pct', None)
> >
> >                      # If the entrance exam box on the settings screen has been checked,
> > -                    # and the course does not already have an entrance exam attached...
> > -                    if entrance_exam_enabled and not course_entrance_exam_present:
> > +                    if entrance_exam_enabled:
> >                          # Load the default minimum score threshold from settings, then try to override it
>
> Hmmm, matt i think i should revert back this condition into its original
> state like as >
>
> if entrance_exam_enabled and not course_entrance_exam_present:
> Because during creating an entrance exam , we are first cleaning out any
> old entrance exams in entrance_exam.py line 84.
> So this condition > if entrance_exam_enabled: -- will create a new
> instance of an entrance exam instead of updating the minimum passing score
> of a pre created exam.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6997/files#r24797292>.
>

cahrens commented:
Here is the TNL ticket for reviewing this PR-- https://openedx.atlassian.net/browse/TNL-1465.

Please add Jasmine test coverage for these lines--
https://build.testeng.edx.org/job/edx-platform-report-pr/1655/Diff_Coverage_Report/

marcotuts commented:
@ziafazal, @mattdrayer  - SOl-221 is the sequential messaging story, so I would expect the text in this sandbox to match the desired final text (cc'ed @srpearce). While reviewing this sandbox I found issues with the text being rendered. http://pr6997.m.sandbox.edx.org/courses/edx/101/101/courseware/52c81ed9df0140b79ad827bee401cde2/7eca3540ec70454a9eef7c41f7e52c7d/

"You have successfully passed the exam. Your current score is 0%."

I'm also not seeing the proper gating happening here prior to completing the exam, not sure if this is a configuration issue?

asadiqbal08 commented:
test this please

mattdrayer commented:
@cahrens -- just a heads up that @asadiqbal08 has addressed the missing test coverage you noted:  https://build.testeng.edx.org/job/edx-platform-all-tests-pr/1808/SHARD=1,TEST_SUITE=unit,label_exp=jenkins-worker/artifact/reports/diff_coverage_combined.html

mattdrayer commented:
@marcotuts @asadiqbal08 -- okay, it took a bit of digging but I was able to determine that the course-milestone requirement relationship had been deactivated on the sandbox -- so all of the meta flags were set correctly and the proper milestone entities were in the database, but one piece of the puzzle was missing.  Pretty sure this is a bug having to do with either editing or (re)enabling an entrance exam for a course, so will have to look a bit closer at it tomorrow.  In the meantime I've manually activated the link so the gating should be behaving correctly now.

marcotuts commented:
@ziafazal, @mattdrayer - While reviewing @srpearce's copy suggestions, a screenshot made it seem like the entrance exam sequential message was being rendered outside of the context of the entrance exam -subsection. This message should not be rendering outside of that subsection. (I think this is the right PR to be commenting on, please let me know if this should be elsewhere). 
![image](https://cloud.githubusercontent.com/assets/2023680/6271329/f9edccac-b82d-11e4-8e9c-4684fda69642.png)

srpearce commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -128,6 +129,14 @@ def _create_entrance_exam(request, course_key, entrance_exam_minimum_score_pct=N
+        parent_locator=unicode(created_block.location),
+        user=request.user,
+        category='sequential',
+        display_name=_('Entrance Exam - Subsection')
```
I'm not sure "Subsection" should be included in the display name

asadiqbal08 replied:
@marcotuts please conform this.

chrisndodge replied:
Seems like this still needs resolution, but something this small shouldn't prevent moving to 2nd review.

marcotuts replied:
@asadiqbal08 - Apologies for the delay here, I thought the title above came out of doc review, and @srpearce I interpreted your comment above as "Subsection should be included in the display name."

My suggestion for this string would be to use "Exam"

Thanks @cahrens for flagging this. I had misread the earlier comments.

srpearce commented on cms/templates/js/course-outline.underscore:
```diff
@@ -116,7 +119,14 @@ if (xblockInfo.get('graded')) {
+            <% } else if (xblockInfo.getExamScore() !=null) { %>
+                <div class="status-exam">
+                    <span>
+                          <%= gettext('You must score ')%> <%= xblockInfo.getExamScore() %> <%= gettext('% or higher to proceed')%>
```
Change to "Students must score x% or higher to access course materials."

asadiqbal08 replied:
@srpearce Done.

asadiqbal08 commented:
@marcotuts I have updated the code. Now message will appear under entrance exam subsection context. Let me know if anything else required. I also updated the text in CMS. 
@mattdrayer

mattdrayer commented:
@asadiqbal08 @marcotuts -- I finally figured out what was going on with the funky gating logic -- basically there was a bit of contention for the course+milestone link when both prerequisite courses and entrance exams were enabled, but no prerequisite courses were specified for the course.  I've made the change and will provision a new sandbox once Jenkins has given the thumbs-up.  @ziafazal, just FYI on the change I made (see commit).

asadiqbal08 commented:
@mattdrayer is sandbox configured now for entrance exam ?

mattdrayer commented:
Yes, it should be available now at the pr6997 URL

On Mon, Feb 23, 2015 at 7:17 AM, Asad <notifications@github.com> wrote:

> @mattdrayer <https://github.com/mattdrayer> is sandbox configured now for
> entrance exam ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6997#issuecomment-75531860>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

asadiqbal08 commented:
test this please

mattdrayer commented:
Jenkins, test this please.

mattdrayer commented:
@chrisndodge -- if you can take a final spin through this merger of #6811 and #6854, that would be great, thanks.  Looks like Jenkins approves of the branch at this point (most recent fail was due to flaky library test)

chrisndodge commented:
OK, I see that this looks like what I reviewed on closed PR's 6811 and 6854. I can't closely review the Studio JS, so I'll rely on 2nd review that has domain expertise on that.

+1 to move onwards down the review pipeline

smagoun commented:
@mattdrayer looks like this needs a rebase?

mattdrayer commented:
Yep, will rebase and then provision a new sandbox
On Mar 2, 2015 8:52 AM, "Steve Magoun" <notifications@github.com> wrote:

> @mattdrayer <https://github.com/mattdrayer> looks like this needs a
> rebase?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6997#issuecomment-76714494>.
>

cahrens commented:
The CSS looks really off on this sandbox.

![image](https://cloud.githubusercontent.com/assets/484484/6446205/7eb00304-c0d7-11e4-8b71-ea8d21fb4f7e.png)

cahrens commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -128,6 +129,14 @@ def _create_entrance_exam(request, course_key, entrance_exam_minimum_score_pct=N
+    create_xblock(
+        parent_locator=unicode(created_block.location),
+        user=request.user,
+        category='sequential',
```
It looks like there is still disagreement about the subsection name (comments below).

My opinion is that the subsection name should be visible on the outline because it is shown in 2 places on the unit page (top breadcrumb and RHS navigator). Since the user will be exposed to the name anyway, hiding it on the outline doesn't do anything and just makes the UI inconsistent.

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -172,13 +172,17 @@ function(Backbone, _, str, ModuleUtils) {
-                //hide/remove the delete trash icon if type is entrance exam.
-                if (_.has(type, 'is_entrance_exam') && type['is_entrance_exam']) {
+            if(type !=null) {
+                //hide the delete/drag icon if type is entrance exam.
```
I disagree with hiding the drag icon on the entrance exam because it is still possible to drag other sections before it.

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -172,13 +172,17 @@ function(Backbone, _, str, ModuleUtils) {
+         * Return true if section content is required to show.
+         * currently, check section for an entrance exam.
+        */
+        showSectionContent: function(){
```
This method name, "showSectionContent" does not make sense-- I don't see how it maps to hiding Delete and Drag.

These should stay as separate methods, and xblocks should indicate their desired behavior so that is_entrance_exam does not need to appear in this file.

For instance, if an xblock should not be deletable, that is the property that should be set instead of "is_entrance_exam".

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -186,10 +190,50 @@ function(Backbone, _, str, ModuleUtils) {
+         * Return true if subsection content is required to show.
+         * currently, check subsection for an entrance exam.
+        */
+        showSubSectionContent: function() {
```
This is a very confusing method because the subsection content (ie, units) are still shown. What is being hidden in the release date, delete icon, etc.

Again, is_entrance_exam_subsection should not be in this file, so the xblock needs to state somehow that it does not want subsection control actions (or some similar name-- I haven't given it much thought).

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -186,10 +190,50 @@ function(Backbone, _, str, ModuleUtils) {
+           return type;
+        },
+
+       getExamScore: function() {
```
It doesn't make sense to have the exam score as part of xblock_info.

mattdrayer replied:
@cahrens -- is there a better-suited location already in place for this logic, or would it be better to add a new file/script?

mattdrayer replied:
For example, I see chapters.js and section.js -- maybe one of these?

cahrens replied:
Similar to my other comments in this file, things that are specific to entrance exams shouldn't be in this file. Instead, general concepts such as "deletable" or "explanatory message" should be used. The desire here is to show a bit of text beneath the Section title similar to the "released date" text. So I would recommend brainstorming how this text could be passed through in a general way, instead of adding a method to this class specifically about an exam score. In fact, perhaps the "released date" text is just an example of this.

cahrens commented:
@mattdrayer I'm going to hold of on reviewing this further until the is_entrance_exam logic is moved out of xblock_info. @andy-armstrong has commented on this in past code reviews, and I don't want more assessment-specific logic pushed into what should be a "type agnostic" class.

marcotuts commented:
@cahrens - Thanks for the feedback on this PR! A number of issues have been raised here (or in hipchat) that I've summarized below:

(cc'ed @mattdrayer)
 
1.) The LMS is still showing the sidebar in the case of a failing / incomplete entrance exam, though that was originally expected to be a part of SOL-199. I've created SOL-448 to deal with this remaining experience change in the future. (https://openedx.atlassian.net/browse/SOL-448) 

2.) During the review process Christina discovered that you can complete two undesired drag and drop operations when Entrance Exams are enabled. The first is that you can drag sections above the entrance exam. The second (and more concerning one) is that you can drag/drop additional subsections into the entrance exams and then not drag them out. Both of these issues will be tackled in a separate follow-on story, with a specific goal of seeing if we can fix the second case sooner rather than later.  (https://openedx.atlassian.net/browse/SOL-447) 

3.) Entrance exams have a message above the sequential that shows your score, but this doesn't update asynchronously, which means users would have to refresh the page to see the latest score. This is also a known limitation and is planned for via (https://openedx.atlassian.net/browse/SOL-222)

For the purposes of this PR's code review, the three issues above can be considered out of scope. I think there are still other outstanding issues about entrance_exam logic (as noted above), but I wanted to make sure we were all on the same page about the other product / ux questions raised. I've also had a chance to review these changes with @explorerleslie.

Thanks everyone!

mattdrayer commented:
@cahrens -- we've gone ahead and addressed the Javascript items you've noted, so the branch should be in better shape now.  I noted one small quality violation which i've repaired and pushed, and Jenkins is currently analyzing the branch.  However, aside from that one issue, the previous build looked good.

cahrens commented:
@mattdrayer I'll start re-reviewing. Note though that this PR needs a rebase.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -781,10 +781,17 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
+        if parent_xblock is None:
+            parent_xblock = get_parent_xblock(xblock)
+
+        explanatory_message = 'Students must score {score}% or higher to access course materials.'.format(
```
i18n

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -805,8 +812,14 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
     }
+
+    # Entrance exam subsection should be hidden.
+    if xblock.category == 'sequential' and getattr(parent_xblock, "is_entrance_exam", False):
```
You don't check here if parent_xblock is already defined, though you do on line 790. What will happen if parent_xblock is None?

I'm a little confused about making the subsection not visible, as I would think that also hides the units. Perhaps the rest of the code will make this clearer.

asadiqbal08 replied:
according to me, parent_xblock should not be none in case of entrance exam when xblock category is sequential' . entrance_exam.py creating entrance exam and its subsection.

cahrens replied:
But it is possible that there is a sequential that is not related to an entrance exam. So is it possible that there is a sequential not related to an entrance exam, and parent_xblock is None? What is the value of getattr(parent_xblock, "is_entrance_exam", False) if parent_xblock is None?

asadiqbal08 replied:
if parent_xblock is none then default will be False.

cahrens commented:
@marcotuts I still feel strongly that we should not hide the subsection name on the course outline because it is still visible (in 2 places) on the unit page. Can I convince you of that?

cahrens commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -1405,20 +1405,6 @@ def test_json_responses(self):
         json_response = json.loads(resp.content)
         self.validate_course_xblock_info(json_response, course_outline=True)
 
-    def test_chapter_entrance_exam_xblock_info(self):
```
You should instead test the new logic you added about deletable, addable, draggable, explanatory_text, and is_visible (please also test default behavior for when the xblock is not related to an entrance exam.

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -133,9 +133,18 @@ function(Backbone, _, str, ModuleUtils) {
+             */
+            'is_visible': null,
+            /**
+             * Explanatory message for xblock that need to show.
```
Suggested rewording, "Optional explanatory message about the xblock."

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -133,9 +133,18 @@ function(Backbone, _, str, ModuleUtils) {
+             * actions defines the state of delete, drag and add functionality for a xblock.
              */
-            'override_type': null
+            'actions': null,
```
List the actual keys that are supported in the documentation.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -781,10 +781,17 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
     if getattr(xblock, "is_entrance_exam", None):
-        override_type['is_entrance_exam'] = xblock.is_entrance_exam
+        # Entrance exam section should not be deletable, draggable and not have 'New Subsection' button.
+        xblock_actions['deletable'] = xblock_actions['addable'] = xblock_actions['draggable'] = False
```
"deletable" and "draggable" refer to the actual xblock, correct? "addable" refers to whether or not children can be added to the xblock. Please rename "addable" to make that clear-- perhaps "childAddable".

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -172,13 +181,33 @@ function(Backbone, _, str, ModuleUtils) {
+        },
+
+        /**
+         * Return true if action is required e.g. delete, drag, add new etc.
```
Define default behavior if the key is not present.

cahrens commented on cms/static/js/spec/models/xblock_info_spec.js:
```diff
@@ -7,16 +7,6 @@ define(['backbone', 'js/models/xblock_info'],
                 expect(new XBlockInfo({'category': 'vertical'}).isEditableOnCourseOutline()).toBe(true);
             });
-
-            it('cannot delete an entrance exam', function(){
```
Add unit test for behavior of isActionRequired.

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -259,6 +267,11 @@ define(["jquery", "sinon", "js/common_helpers/ajax_helpers", "js/views/utils/vie
                     expect(getItemsOfType('unit')).not.toExist();
                 });
+
+                it('unit initially exist for entrance exam', function() {
```
Test that subsection does not exist.

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -86,18 +86,27 @@ if (xblockInfo.get('graded')) {
         <% if (!xblockInfo.isVertical()) { %>
-            <div class="status-release">
+            <% if (xblockInfo.get('explanatory_message') !=null) { %>
+                <div class="status-exam">
```
This class name should be about the explanatory message, not "status-exam".

cahrens commented on cms/static/js/views/xblock_outline.js:
```diff
@@ -44,6 +44,19 @@ define(["jquery", "underscore", "gettext", "js/views/baseview", "js/views/utils/
                 this.addButtonActions(this.$el);
                 this.addNameEditor();
+
+                if (this.model.get('is_visible') != null) {
```
Add test coverage of this.

Subsections are expanded by default, so I'm a little surprised that the locators are not already present. Or perhaps they are, but you are being complete in case is_visible is ever specified on a Section? I can support that, but I'm curious about how this interacts with the default expanded state for subsections.

andy-armstrong replied:
Change ``!=`` to ``!==`` as that is our best practice.

cahrens commented on common/test/acceptance/tests/studio/test_studio_settings_details.py:
```diff
@@ -161,3 +161,32 @@ def test_enable_entrance_exam_for_course(self):
+
+    def test_entrance_exam_has_unit_button(self):
+        """
+        Test that entrance exam should be created after checking the 'enable entrance exam' checkbox.
```
Going forward on other projects, please follow the documentation best practices for bok choy tests, listed on

https://openedx.atlassian.net/wiki/display/TNL/Code+Review+Guidelines+and+Gotchas (see bok choy section).

cahrens commented on lms/templates/courseware/courseware.html:
```diff
@@ -234,6 +234,25 @@
     </div>
 % endif
     <section class="course-content" id="course-content">
+        % if entrance_exam_current_score is not UNDEFINED and \
```
I'm not familiar with UNDEFINED in HTML templates. Where is that defined?

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -805,8 +812,14 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
+
+    # Entrance exam subsection should be hidden.
+    if xblock.category == 'sequential' and getattr(parent_xblock, "is_entrance_exam", False):
+        xblock_info["is_visible"] = False
```
Instead of "is_visible", @andy-armstrong and I prefer "is_header_visible". We both think that "is_visible" would refer to the children as well, and it is a general concept that we may wish to support in the future (thus we would like to reserve the property name "is_visible").

asadiqbal08 commented:
@mattdrayer today i addressed the maximum suggestions from @cahrens . Thanks @cahrens for the code review. I rebased it with master But I need to add some JS tests for missing coverage. I will do it tomorrow.

asadiqbal08 commented:
@mattdrayer @cahrens I have added the missing JS tests. kindly let me know if anything else required.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -805,8 +812,18 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
+        "explanatory_message": explanatory_message
     }
+
+    # Entrance exam subsection should be hidden. in_entrance_exam is inherited metadata, all children will have it.
```
If I am reading your comment correctly, you are saying that if the sequential has "in_entrance_exam" set to True, then its parent will always have "in_entrance_exam" set to True, due to the inheritance. Then why are you also checking the parent status? Can you just delete lines 821-824?

asadiqbal08 replied:
@cahrens let me explain what i am doing here: 
1) hiding the header contents of subsection('sequential') for an entrance exam (not for verticals/units). 
2) is_entrance_exam is set True for chapter (not for subsections and nor for verticals). 
3) in_entrance_exam is True for all children even for parent too (you are right) But here we are interested to know that xblock is subsection ('sequential') of entrance exam ('chapter') but NOT for verticals/units so that is why i am checking that the parent of xblock ('sequential') is entrance exam.

cahrens replied:
I am still confused why lines 821-824 are necessary. I understand you want to check specifically for xblock.category == 'sequential', and I agree that should stay. But if the condition "xblock.category == 'sequential' and getattr(xblock, "in_entrance_exam", False)" is True, won't it always be the case that if getattr(parent_xblock, "is_entrance_exam", None) is also true? To state it another way, is there any case where you can have a sequential with in_entrance_exam == True, but its parent has in_entrance_exam == False?

mattdrayer replied:
@asadiqbal08 -- I've left this one for you to work through -- please let me know if you need a hand.

asadiqbal08 replied:
Ok

asadiqbal08 replied:
done

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -781,10 +781,17 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
+    # defining the default value 'True' for delete, drag and add new child actions in xblock_actions for each xblock.
+    xblock_actions = {'deletable': True, 'draggable': True, 'childAddable': True}
+    explanatory_message = None
     if getattr(xblock, "is_entrance_exam", None):
```
Should this logic be specific to xblock.category == 'chapter' (similar to block below that is just for sequentials)?

mattdrayer replied:
@asadiqbal08 -- I've left this one as an exercise for you -- let me know if you need a hand.

cahrens commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -1416,9 +1416,68 @@ def test_chapter_entrance_exam_xblock_info(self):
+        actions = xblock_info['actions']
+        self.assertEqual(actions['deletable'], False)
+        self.assertEqual(actions['draggable'], False)
+        self.assertEqual(actions['childAddable'], False)
```
Add check that self.assertIsNone(xblock_info.get('is_header_visible', None))

mattdrayer replied:
Added

cahrens commented on cms/static/js/models/xblock_info.js:
```diff
@@ -172,13 +182,33 @@ function(Backbone, _, str, ModuleUtils) {
+        },
+
+        isHeaderVisible: function(){
+            if(this.get('is_header_visible')!= null) {
```
For JavaScript, use !== and === instead of != and == (so that it does not do type coercion). If you are not using an editor that does JS linting, I would recommend that you change to one that does, or investigate how to run jslint in edx-platform (it was added by some developers who are no longer here, and I'm not sure how to run it).

mattdrayer replied:
I think I've updated the offending spots here -- there were only a couple from what I could see -- @asadiqbal08 FYI

cahrens commented on cms/static/js/spec/models/xblock_info_spec.js:
```diff
@@ -7,16 +7,38 @@ define(['backbone', 'js/models/xblock_info'],
-            it('cannot delete an entrance exam', function(){
-                expect(new XBlockInfo({'category': 'chapter', 'override_type': {'is_entrance_exam':true}})
-                    .canBeDeleted()).toBe(false);
+            it('works correct to show icons e.g. trash icon, drag when actions are required', function(){
```
Please also test the default behavior (when each of these are not specified, the helper "is" methods should still return true).

mattdrayer replied:
I've added a set of tests to cover each one of these methods when "actions" is empty

cahrens commented on cms/static/js/views/xblock_outline.js:
```diff
@@ -44,6 +44,19 @@ define(["jquery", "underscore", "gettext", "js/views/baseview", "js/views/utils/
                 this.addButtonActions(this.$el);
                 this.addNameEditor();
+
+                if (this.model.get('is_header_visible') !== null) {
```
Please combine lines 48 and 51 into a single iff statement.

mattdrayer replied:
Done

cahrens commented on cms/static/js/views/xblock_outline.js:
```diff
@@ -44,6 +44,19 @@ define(["jquery", "underscore", "gettext", "js/views/baseview", "js/views/utils/
+
+                if (this.model.get('is_header_visible') !== null) {
+                    // Currently an entrance exam does not contains subsection.
+                    // So we are setting the locator in order to show expanded.
```
This comment, "Currently an entrance exam does not contains subsection." is confusing. It does contain a subsection, but the header controls are hidden, correct? Please update the comment.

mattdrayer replied:
I've reworded this comment

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -150,6 +152,42 @@ def fulfill_course_milestone(course_key, user):
             add_user_milestone({'id': user.id}, milestone)
 
 
+def get_required_content(course, user):
```
This code just moved from lms/djangoapps/courseware/module_render.py, correct?

asadiqbal08 replied:
correct

cahrens commented on lms/djangoapps/courseware/courses.py:
```diff
@@ -441,3 +445,47 @@ def get_problems_in_section(section):
     return problem_descriptors
+
+
+def get_entrance_exam_score(request, course):
```
Where is the unit test coverage for this method?

mattdrayer replied:
@asadiqbal08 -- please add a test to cover this method directly -- it is currently only being covered through the views tests.

asadiqbal08 replied:
@mattdrayer i have added unit tests for this method. please check.

cahrens commented on lms/djangoapps/courseware/courses.py:
```diff
@@ -441,3 +445,47 @@ def get_problems_in_section(section):
+    return calculate_entrance_exam_score(request.user, course, exam_modules)
+
+
+def get_entrance_exam_content_info(request, course):
```
Where is the unit test coverage for this method?

mattdrayer replied:
@asadiqbal08 -- if you can add some additional tests for this method and others, that would be great.

asadiqbal08 replied:
@mattdrayer i have added unit test for this method. please check.

cahrens commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -171,7 +136,9 @@ def toc_for_course(request, course, active_chapter, active_section, field_data_c
         # Check to see if the course is gated on required content (such as an Entrance Exam)
-        required_content = _get_required_content(course, request.user)
+        required_content = []
+        if settings.FEATURES.get('MILESTONES_APP', False):
```
Why was a check of settings.FEATURES.get('MILESTONES_APP', False) introduced? A portion of get_required_content also checks settings.FEATURES.get('MILESTONES_APP', False).

asadiqbal08 replied:
@mattdrayer

mattdrayer replied:
I've cleaned up this spot

cahrens commented on lms/djangoapps/courseware/tests/test_entrance_exam.py:
```diff
@@ -206,6 +211,100 @@ def setUp(self):
+            }
+        )
+        resp = self.client.get(url)
+        self.assertEqual(resp.status_code, 200)
```
Test that "Exam Problem - Problem 1" and "Exam Problem - Problem 2" are present on the page. Currently there is no test point that ever verifies they are present when they should be present-- there is only a test point that verifies that they are absent in one case. That means that test_entrance_exam_content_absence could be passing incorrectly if the problems were never added with those exact names to the course.

mattdrayer replied:
I added a new test to cover this use case but noticed that it was failing.  Unfortunately I'm out of time today but perhaps @asadiqbal08 can pick it up where I left off.

cahrens commented on lms/djangoapps/courseware/tests/test_entrance_exam.py:
```diff
@@ -206,6 +211,100 @@ def setUp(self):
         )
 
+    @mock.patch('xmodule.x_module.XModuleMixin.has_dynamic_children', mock.Mock(return_value='True'))
+    def test_view_redirect_if_entrance_exam_required(self):
```
@mattdrayer Should this PR prevent you from directly hitting a portion of the course that is "beyond" the entrance exam? On the sandbox, I can go directly to http://pr6997.m.sandbox.edx.org/courses/wew/wefe/wewe/courseware/cad30653cd2b44bcafde2d34a959897b/618fd05fe56347d4808a74d3b7ae1936/1 (logged in as verified@example.com), even though I have not finished the entrance exam.

mattdrayer replied:
@cahrens -- the gating logic is designed to unlock the rest of the course after you have acheived the minimum score required -- so if the minimum score is 50%, and there are two problems in the exam, and you answer one correctly, the rest of the course is now available to you. @marcotuts -- that is how it's supposed to work, correct?

cahrens replied:
Sorry, I meant to say "PASSED" the entrance exam. I have not passed it-- "To access course materials, you must score 80% or higher on this exam. Your current score is 0%."

mattdrayer replied:
@cahrens -- by 'going directly' can you explain further what you are experiencing?  I've logged in as the Verified user on the sandbox and I can't seem to get around the entrance exam gating.  In addition, the messaging indicates that the required score is 100% and my score is currently 0%.

cahrens replied:
I directly hit the URL for a section outside of the entrance exam.

As far as the score, if you go to http://studio.pr6997.m.sandbox.edx.org/settings/details/wew/wefe/wewe you will see that the exam score should be 80% at the moment, not 100%. I went back and forth a couple times before I wrote up the bug, so the current state does not match the repro steps.

mattdrayer replied:
Yeah, I tried that as well -- ie, hitting the URL you provided directly --
after logging in I was redirected to the entrance exam page/content -- will
keep poking at it

On Fri, Mar 6, 2015 at 1:55 PM, Christina Roberts <notifications@github.com>
wrote:

> In lms/djangoapps/courseware/tests/test_entrance_exam.py
> <https://github.com/edx/edx-platform/pull/6997#discussion_r25967637>:
>
> > @@ -206,6 +211,100 @@ def setUp(self):
> >              ]
> >          )
> >
> > +    @mock.patch('xmodule.x_module.XModuleMixin.has_dynamic_children', mock.Mock(return_value='True'))
> > +    def test_view_redirect_if_entrance_exam_required(self):
>
> I directly hit the URL for a section outside of the entrance exam.
>
> As far as the score, if you go to
> http://studio.pr6997.m.sandbox.edx.org/settings/details/wew/wefe/wewe you
> will see that the exam score should be 80% at the moment, not 100%. I went
> back and forth a couple times before I wrote up the bug, so the current
> state does not match the repro steps.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6997/files#r25967637>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

mattdrayer replied:
I've put a fix in place for the score not being updated -- let's see how it works on a new sandbox.  @asadiqbal08, if you can take a closer look at the gating logic that would be great -- I was never able to get around it, even using the direct link, but maybe it has to do with some sort of built-up (or invalid) state on the server.

asadiqbal08 replied:
@mattdrayer i think @ziafazal have to add some gating logic. there are some other tickets SOL-216/219.

mattdrayer commented:
@cahrens -- the gating logic is designed to unlock the rest of the course after you have acheived the minimum score required -- so if the minimum score is 50%, and there are two problems in the exam, and you answer one correctly, the rest of the course is now available to you.  @marcotuts -- that is how it's supposed to work, correct?

cahrens commented on lms/djangoapps/courseware/views.py:
```diff
@@ -412,6 +413,19 @@ def _index_bulk_op(request, course_key, chapter, section, position):
+                    return redirect('courseware_section',
+                                    course_id=course_key.to_deprecated_string(),
+                                    chapter=exam_chapter.url_name,
+                                    section=exam_section.url_name)
```
This will error if exam_section is None (which will be the case if exam_chapter.get_children() returns 0).

mattdrayer replied:
@asadiqbal08 -- FYI, I've addressed this issue

cahrens commented on lms/templates/courseware/courseware.html:
```diff
@@ -234,6 +234,25 @@
     </div>
 % endif
     <section class="course-content" id="course-content">
```
I am still curious to know how UNDEFINED works. Can you please explain it to me?

mattdrayer replied:
@cahrens -- did a bit of research into this one -- UNDEFINED is the default value for a missing parameter on a mako template -- see here: http://docs.makotemplates.org/en/latest/runtime.html#context-variables

cahrens replied:
Thanks, good to know!

cahrens commented:
@marcotuts @mattdrayer A couple of things that I would consider serious bugs.

1) I enabled the entrance exam in my course with a required passing score of 80%. I added a problem to the exam and published it. When I view the course in the LMS, I correctly get the message that the required score is 80%.

2) I then changed the required score in Studio to 100%. The message in LMS (and Studio, on the outline page) will not update, no matter how many refreshes and such, to say that the required score is 100% instead of 80%. I do not know what would happen if a user got 90% correct-- would they pass or not (as a side note, for students who have already passed the exam, what happens if the required score changes)?

3) It seems that the only way I can get the LMS and Studio to update with the new required score is to disable the entrance exam and add it again. But that deletes the content I had in the entrance exam, which is certainly unexpected and annoying.

mattdrayer commented:
@cahrens -- I've made some changes based on your feedback.  There are still some items for @asadiqbal08 to update, however.   Also, I think I've gotten to the bottom of the score updating issue and have added a fix for that -- it seems to be working correctly on devstack, so I've pushed the changes and will provision a new sandbox over the weekend for us to look at.

cahrens commented on cms/static/js/views/xblock_outline.js:
```diff
@@ -44,6 +44,18 @@ define(["jquery", "underscore", "gettext", "js/views/baseview", "js/views/utils/
+                // For cases in which we need to suppress the header controls during rendering, we'll
+                // need to add the current model's id/locator to the set of expanded locators
+                if (this.model.get('is_header_visible') !== null) && (!this.model.get('is_header_visible')) {
+                        var locator = this.model.get('id');
```
Fix the spacing.

asadiqbal08 commented:
one thing to note, updating exam score is not the part of SOL-221 and SOL-199 but handling the score update scenario in this PR is great! thanks @mattdrayer

cahrens commented:
@mattdrayer let me know when the sandbox is updated and this is ready for (hopefully final) review.

mattdrayer commented:
@cahrens -- the new sandbox is now ready -- same URL (pr6997) -- let me know if you run into any issues -- thanks!

cahrens commented:
@mattdrayer There is one edge case you might want to consider.

If the user has started the entrance exam, but not passed it, changes to the required score are displayed, but do not take effect.

Repro steps:
1) Create course with entrance exam required as 80%.
2) Student answers some questions on exam, does not yet have 80%.
3) Course changed to only require 50%.
4) Student seems message (on refresh) that 50% is required, but student still has to score 80% to pass.

I certainly agree with not changing the requirement for students who have already passed the exam, but this edge case doesn't seem quite right.

cahrens commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -123,11 +136,19 @@ def _create_entrance_exam(request, course_key, entrance_exam_minimum_score_pct=N
+        category='sequential',
+        display_name=_('Entrance Exam - Subsection')
+    )
+
```
Don't forget to update the subsection name per Marco's comment on line 149.

cahrens commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -181,6 +202,22 @@ def _get_entrance_exam(request, course_key):  # pylint: disable=W0613
+    metadata = exam_data
+    CourseMetadata.update_from_dict(metadata, course, request.user)
+
+    return HttpResponse(status=200)
```
This method returns values, but I don't think it should (the return values are ignored by the only place that calls the method).

cahrens commented on lms/djangoapps/courseware/tests/test_entrance_exam.py:
```diff
@@ -206,6 +228,186 @@ def setUp(self):
+                               })
+        resp = self.client.get(url)
+        self.assertRedirects(resp, expected_url, status_code=302, target_status_code=200)
+        resp = self.client.get(expected_url)
```
You can just add these 3 lines (279-281) to test_view_redirect_if_entrance_exam_required and delete this test case. I believe it is redundant with test_view_redirect_if_entrance_exam_required, though I don't understand why test_view_redirect_if_entrance_exam_required requires a mock of xmodule.x_module.XModuleMixin.has_dynamic_children and also does not need to specify {'ENTRANCE_EXAMS': True}.

Please just take a look at these 2 test cases and see if it makes sense to consolidate them.

asadiqbal08 replied:
done

cahrens commented:
:+1: A couple minor things that need to be cleaned up (that I don't need to re-review), but otherwise looks good.

mattdrayer commented:
Thanks, Christina!!!  I'll address these remaining issues tonight and
squash the commits -- once I have the green light from Jenkins we'll merge
to master.

On Mon, Mar 9, 2015 at 4:23 PM, Christina Roberts <notifications@github.com>
wrote:

> [image: :+1:] A couple minor things that need to be cleaned up (that I
> don't need to re-review), but otherwise looks good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6997#issuecomment-77934151>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

mattdrayer commented:
WOW, we are merging this one!
